trieste/ask_tell_optimization.py

```diff
@@ -0,0 +1,181 @@
+    """
+
+    def __init__(
+        self,
+        search_space: SP,
+        datasets: Mapping[str, Dataset] | Dataset,
+        model_specs: Mapping[str, ModelSpec] | ModelSpec,
+        fit_model: bool,
+        acquisition_rule: AcquisitionRule[S, SP],
+        acquisition_state: Optional[S] = None,
+    ):
+        """
+        :param search_space: The space over which to search.
+        :param datasets: The known observer query points and observations for each tag.
+        :param model_specs: The model to use for each :class:`~trieste.data.Dataset` in
+            ``datasets``.
+        :param fit_model: If `True`, models passed in will be optimized on the given data.
+            If `False`, the models are assumed to be optimized already.
+        :param acquisition_rule: The acquisition rule, which defines how to search for a new point
+            on each optimization step.
+        :param acquisition_state: The optional acquisition state for stateful acquisitions.
+        :raise ValueError: If any of the following are true:
+            - the keys in ``datasets`` and ``model_specs`` do not match
+            - ``datasets`` or ``model_specs`` are empty
+        """
+        self._search_space = search_space
+        self._acquisition_rule = acquisition_rule
+        self._acquisition_state = acquisition_state
+
+        if not datasets or not model_specs:
+            raise ValueError("dicts of datasets and model_specs must be populated.")
+
+        if isinstance(datasets, Dataset):
+            datasets = {OBJECTIVE: datasets}
+            # ignore below is due to MyPy not being able to handle overloads properly
+            model_specs = {OBJECTIVE: model_specs}  # type: ignore
+
+        # reassure the type checker that everything is tagged
+        datasets = cast(Dict[str, Dataset], datasets)
+        model_specs = cast(Dict[str, ModelSpec], model_specs)
+
+        if datasets.keys() != model_specs.keys():
+            raise ValueError(
+                f"datasets and model_specs should contain the same keys. Got {datasets.keys()} and"
+                f" {model_specs.keys()} respectively."
+            )
+
+        self._datasets = datasets
+        self._models = map_values(create_model, model_specs)
+
+        if fit_model:
+            # there is no default value for `fit_model` parameter
+            # to force the user to think about it
+            # before invoking ask/tell API
+            # there is no sensible default here:
+            # `False` may result in suboptimal performance
+            # `True` may result in two model training calls per optimization step
+            for tag, model in self._models.items():
+                dataset = datasets[tag]
+                model.update(dataset)
+                model.optimize(dataset)
+
+    @classmethod
+    def from_record(
+        cls, search_space: SP, acquisition_rule: AcquisitionRule[S, SP], record: Record[S]
```

Comment:
Maybe put record first and make acquisition_rule optional (assuming you end up doing that for init)? Also it might make sense to expose search_space and acquisition_rule as @properties. (There'd probably be no harm doing that for `BayesianOptimizer` too, though it's more obviously useful here.)

Reply:
yup, now that the rule can be optional this needs some reordering indeed

Reply:
what would properties be good for?